### PR TITLE
SmugmugAuthServiceExtension

### PR DESCRIPTION
**Pull request by seehamrun**
This uses the Scribe API, which is the same API that is used in the Flickr4Java library that we pulled in for flickrAuth (and is modeled heavily off of the Flickr4Java api). Note that this does not enable the transfer of Smugmug data as the auth data is not plumbed through or used in the TransferExtension yet. There's still a lot of work left to enable auth in the transfer importer and exporter. 

#314

### CODE REVIEW

File: extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/AuthInterface.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import com.google.common.base.Strings;
+import java.io.IOException;
+import org.dataportabilityproject.types.transfer.auth.TokenSecretAuthData;
+import org.scribe.builder.ServiceBuilder;
+import org.scribe.model.Token;
+import org.scribe.model.Verifier;
+import org.scribe.oauth.OAuthService;
+
+public class AuthInterface {
+
```
Comment (rtannenbaum) [inline extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/AuthInterface.java]: nit: extra newline

File: extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/AuthInterface.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+package org.dataportabilityproject.auth.smugmug.SmugMugAuth;
+
+import com.google.common.base.Strings;
+import java.io.IOException;
+import org.dataportabilityproject.types.transfer.auth.TokenSecretAuthData;
+import org.scribe.builder.ServiceBuilder;
+import org.scribe.model.Token;
+import org.scribe.model.Verifier;
+import org.scribe.oauth.OAuthService;
+
```
Comment (rtannenbaum) [inline extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/AuthInterface.java]: nit: public class should have javadoc (seen in at least a couple files)

File: extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/AuthInterface.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  private final String apiKey;
+  private final String apiSecret;
+
+  public AuthInterface(String apiKey, String apiSecret) {
+    this.apiKey = apiKey;
+    this.apiSecret = apiSecret;
+  }
+
+  public TokenSecretAuthData getRequestToken(String callbackUrl) throws IOException {
+    String callback = (Strings.isNullOrEmpty(callbackUrl)) ? "oob" : callbackUrl;
```
Comment (rtannenbaum) [inline extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/AuthInterface.java]: what's oob? :confused:

File: extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/SmugmugApi.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+package org.dataportabilityproject.auth.smugmug.SmugMugAuth;
+
+import org.scribe.builder.api.DefaultApi10a;
+import org.scribe.model.Token;
+
+/* Smugmug API to use with Scribe Service Builder
+ * URLs provided from the Oauth1.0a flow described here:
+ * https://api.smugmug.com/api/v2/doc/tutorial/authorization.html
+ */
+public class SmugmugApi extends DefaultApi10a {
```
Comment (rtannenbaum) [inline extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/SmugmugApi.java]: nit: how about SmugmutOauthApi?

File: extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuthServiceExtension.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import org.dataportabilityproject.spi.api.auth.AuthServiceProviderRegistry.AuthMode;
+import org.dataportabilityproject.spi.api.auth.extension.AuthServiceExtension;
+import org.dataportabilityproject.spi.cloud.storage.AppCredentialStore;
+import org.dataportabilityproject.types.transfer.auth.AppCredentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SmugMugAuthServiceExtension implements AuthServiceExtension {
+
+  private static final String SMUGMUG_KEY = "SMUGMUG_KEY"; // TODO: Possibly SMUGMUG_API_KEY?
```
Comment (rtannenbaum) [inline extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuthServiceExtension.java]: can you explain the TODO? (or is it still needed?)

File: extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/AuthInterface.java (hunk 2 above)
Comment (seehamrun) [reply extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/AuthInterface.java]: Done.

File: extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/AuthInterface.java (hunk 1 above)
Comment (seehamrun) [reply extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/AuthInterface.java]: Done.

File: extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/AuthInterface.java (hunk 3 above)
Comment (seehamrun) [reply extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/AuthInterface.java]: Done. moved to static field and added comment.

File: extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/SmugmugApi.java (hunk 4 above)
Comment (seehamrun) [reply extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuth/SmugmugApi.java]: Done.

File: extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuthServiceExtension.java (hunk 5 above)
Comment (seehamrun) [reply extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuthServiceExtension.java]: Removed.

File: extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuthDataGenerator.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+  @Override
+  public AuthFlowConfiguration generateConfiguration(String callbackBaseUrl, String id) {
+    // Generate a request token and include that as initial auth data
+    TokenSecretAuthData authData = null;
+    try {
+      authData = smugMugOauthInterface.getRequestToken(callbackBaseUrl + "/callback1/smugmug");
+    } catch (IOException e) {
+      logger.debug("Couldnt get authData {}", e.getMessage());
+      return null;
```
Comment (holachuy) [inline extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuthDataGenerator.java]: Should we just rethrow

File: extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuthServiceExtension.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+  @Override
+  public List<String> getExportTypes() {
+    return supportedServices;
+  }
+
+  @Override
+  public void initialize(ExtensionContext context) {
+    if (initialized) {
+      return;
```
Comment (holachuy) [inline extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuthServiceExtension.java]: Let's log a warning if we try to initialize and already initlalized context

Comment (seehamrun) [reply extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuthServiceExtension.java]: Done.

File: extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuthDataGenerator.java (hunk 6 above)
Comment (seehamrun) [reply extensions/auth/portability-auth-smugmug/src/main/java/org/dataportabilityproject/auth/smugmug/SmugMugAuthDataGenerator.java]: Hm I wanted to be consistent with FlickrAuthDataGenerator and RTM AuthDataGenerator, but it looks like those two aren't consistent with instagram, microsoft or google auth data generators. 

I can create a separate PR To throw a new Runtime or IllegalArg exception but this probably also has to do with the discussion in #176 .

**Pull request merged**